File: frontend/src/types/platform.ts
/**
 * Frontend-owned types for the Platform Administration module
 * (`/api/v1/platform/*`) — defined independently of the backend's
 * Pydantic schemas, matching the convention every other `types/*.ts`
 * file in this project follows. Distinct from `types/admin.ts`
 * (`/api/v1/admin/*`, gated on `role === "admin"`): everything here is
 * gated on `Profile.is_platform_admin` instead, an orthogonal capability
 * — see that field's own doc comment in `types/profile.ts`.
 */

import type { TrendPoint } from "@/types/analytics";

export interface Company {
  id: string;
  name: string;
  slug: string;
  is_active: boolean;
  contact_email: string | null;
  notes: string | null;
  version: number;
  created_at: string;
  updated_at: string;
  deleted_at: string | null;
  deleted_by: string | null;
  is_deleted: boolean;
}

export interface CreateCompanyBody {
  name: string;
}

/** Body for `PATCH /platform/companies/{id}` — setting `is_active: false`
 * suspends every user in the company on their next request; setting it
 * `true` reactivates. Suspending/deleting a platform admin's own company
 * is rejected server-side with a 422 (self-lockout guard). */
export interface UpdateCompanyBody {
  expected_version: number;
  name?: string;
  is_active?: boolean;
  contact_email?: string;
  notes?: string;
}

/** A company's license/plan information, plus computed, informational
 * fields (`seats_used`/`is_expired`) — nothing in the backend enforces
 * these; they exist purely for a platform admin's own reference. */
export interface CompanyLicense {
  company_id: string;
  plan_tier: string;
  seat_limit: number | null;
  expires_at: string | null;
  notes: string | null;
  seats_used: number;
  is_expired: boolean;
  updated_at: string;
}

/** Body for `PATCH /platform/companies/{id}/license` — at least one
 * field required (422 otherwise); works for both first-time creation
 * (when the company has no license yet) and updates. */
export interface CompanyLicenseUpdateBody {
  plan_tier?: string;
  seat_limit?: number | null;
  expires_at?: string | null;
  notes?: string;
}

export interface FeatureFlag {
  key: string;
  description: string;
  enabled: boolean;
  updated_at: string;
}

export interface CreateFeatureFlagBody {
  key: string;
  description: string;
  enabled?: boolean;
}

/** Body for `PATCH /platform/feature-flags/{key}` — at least one field required. */
export interface UpdateFeatureFlagBody {
  description?: string;
  enabled?: boolean;
}

export interface PlatformStats {
  total_tenants: number;
  active_tenants: number;
  total_users: number;
  total_requests: number;
  /** Daily request-creation counts over the trailing 30 days — same
   * `TrendPoint` shape `TrendChart`/analytics already consume, so no new
   * chart primitive is needed for this panel. */
  request_volume_trend: TrendPoint[];
  active_workflow_definitions: number;
  total_storage_bytes: number;
}

/** `redis`/`job_queue`/`dead_letter_by_queue` are only present when Redis
 * is configured on the backend instance — render an "inactive"/"not
 * configured" state when absent, the same convention
 * `QueueStatsPanel`/`/admin/jobs` already uses for its own Redis-optional
 * fields. */
export interface PlatformHealth {
  status: "ok" | "degraded";
  database: "ok" | "unreachable";
  scheduler_active: boolean;
  redis?: "ok" | "unreachable";
  job_queue?: "ok" | "unreachable";
  dead_letter_by_queue?: Record<string, number>;
}

/** `action` intentionally isn't narrowed to a fixed string-literal union
 * (unlike `types/audit.ts`'s `AuditAction`): this feed spans every action
 * code the whole system produces (ordinary request/approval events *and*
 * the new `COMPANY_*`/`FEATURE_FLAG_*` platform events), and new codes
 * may be added over time — a plain `string` with an unrecognized-value
 * fallback in the label/icon maps is more robust than an exhaustive union
 * that breaks the build every time the backend enum grows. */
export interface PlatformAuditEntry {
  id: string;
  actor_id: string | null;
  company_id: string | null;
  request_id: string | null;
  action: string;
  metadata: Record<string, unknown> | null;
  created_at: string;
}

export interface PlatformAuditLogFilters {
  company_id?: string;
  actor_id?: string;
  action?: string;
  created_after?: string;
  created_before?: string;
  page?: number;
  page_size?: number;
}

File: app/api/schemas/platform.py
"""HTTP schemas for the ``platform`` (company management) resource.

``app.models.company.Company`` is already a Pydantic model — ``CompanyOut``
below is a thin ``from_attributes=True`` wrapper around it, the same
"reuse, don't re-derive" technique used throughout this API layer.
"""

from __future__ import annotations

from typing import Any
from uuid import UUID

from pydantic import ConfigDict, Field

from app.models.base import EAHBaseModel, UTCDatetime
from app.models.enums import AuditAction

__all__ = [
    "CompanyOut",
    "CreateCompanyBody",
    "UpdateCompanyBody",
    "RestoreCompanyBody",
    "CompanyLicenseOut",
    "FeatureFlagOut",
    "FeatureFlagCreateBody",
    "TrendPointOut",
    "PlatformStatsOut",
    "PlatformAuditEntryOut",
]


class CompanyOut(EAHBaseModel):
    """Wraps ``app.models.company.Company``."""

    model_config = ConfigDict(from_attributes=True, extra="forbid", frozen=True)

    id: UUID
    name: str
    slug: str
    is_active: bool
    contact_email: str | None
    notes: str | None
    version: int
    created_at: UTCDatetime
    updated_at: UTCDatetime
    deleted_at: UTCDatetime | None
    deleted_by: UUID | None
    is_deleted: bool


class CreateCompanyBody(EAHBaseModel):
    """Body for ``POST /platform/companies``."""

    name: str = Field(min_length=1, max_length=200)


class UpdateCompanyBody(EAHBaseModel):
    """Body for ``PATCH /platform/companies/{id}``."""

    expected_version: int = Field(ge=1)
    name: str | None = Field(default=None, min_length=1, max_length=200)
    is_active: bool | None = None
    contact_email: str | None = None
    notes: str | None = None


class RestoreCompanyBody(EAHBaseModel):
    """Body for ``POST /platform/companies/{id}/restore``."""

    expected_version: int = Field(ge=1)


class CompanyLicenseOut(EAHBaseModel):
    """Wraps ``app.models.company_license.CompanyLicense``."""

    model_config = ConfigDict(from_attributes=True, extra="forbid", frozen=True)

    company_id: UUID
    plan_tier: str
    seat_limit: int | None
    expires_at: UTCDatetime | None
    notes: str | None
    seats_used: int
    is_expired: bool
    updated_at: UTCDatetime


class FeatureFlagOut(EAHBaseModel):
    """Wraps ``app.models.feature_flag.FeatureFlag``."""

    model_config = ConfigDict(from_attributes=True, extra="forbid", frozen=True)

    key: str
    description: str
    enabled: bool
    updated_at: UTCDatetime


class FeatureFlagCreateBody(EAHBaseModel):
    """Body for ``POST /platform/feature-flags``."""

    key: str = Field(min_length=1, max_length=100)
    description: str = Field(min_length=1)
    enabled: bool = False


class TrendPointOut(EAHBaseModel):
    """Wraps ``app.analytics.dto.TrendPoint``."""

    model_config = ConfigDict(from_attributes=True, extra="forbid", frozen=True)

    label: str
    period_start: UTCDatetime
    period_end: UTCDatetime
    value: float


class PlatformStatsOut(EAHBaseModel):
    """Wraps ``app.models.platform_stats.PlatformStats``."""

    model_config = ConfigDict(from_attributes=True, extra="forbid", frozen=True)

    total_tenants: int
    active_tenants: int
    total_users: int
    total_requests: int
    request_volume_trend: list[TrendPointOut]
    active_workflow_definitions: int
    total_storage_bytes: int


class PlatformAuditEntryOut(EAHBaseModel):
    """Wraps ``app.database.repositories.audit_repository.AuditLogRecord``
    for the platform-wide audit history/activity timeline — distinct from
    ``app.api.schemas.audit.AuditEntryOut`` (which wraps a per-request,
    actor-name-resolved read model instead), since this view is
    cross-tenant and shows the raw ``company_id``/``actor_id`` rather
    than a resolved name (the frontend already holds the company list
    client-side and can join locally).
    """

    model_config = ConfigDict(from_attributes=True, extra="forbid", frozen=True)

    id: UUID
    actor_id: UUID | None
    company_id: UUID | None
    request_id: UUID | None
    action: AuditAction
    metadata: dict[str, Any] | None
    created_at: UTCDatetime
